<h2>Data Science Capstone<h3>Part 3 - import data of neighborhoods from Part 2, explore and cluster the data</h3>

In [1]:
import pandas as pd

In [3]:
df_tn = pd.read_csv('TNP2_out.csv')
df_tn.head()

Unnamed: 0 PostalCode      Borough                            Neighborhood  \
0           0        M1B  Scarborough                          Rouge, Malvern   
1           1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union   
2           2        M1E  Scarborough       Guildwood, Morningside, West Hill   
3           3        M1G  Scarborough                                  Woburn   
4           4        M1H  Scarborough                               Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [7]:
df_tn.drop(columns='Unnamed: 0', inplace=True)

In [8]:
df_tn.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [9]:
df_tn.shape

(103, 5)

In [14]:
df_tn['Borough'].unique()

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       "Queen's Park", 'Mississauga', 'Etobicoke'], dtype=object)

<h4>Create a more concise dataframe that includes only those boroughs that have toronto in their names</h4>

In [46]:
df_tb = df_tn.loc[(df_tn['Borough'].isin(['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']))]

In [47]:
df_tb.shape

(38, 5)

In [48]:
df_tb.head()

PostalCode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [49]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_tb['Borough'].unique()),
        df_tb.shape[0]
    )
)

The dataframe has 4 boroughs and 38 neighborhoods.


In [60]:
df_tb.reset_index(drop=True)

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      The Beaches West, India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        Cabbagetown, St. James Town  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Harbourfront, Regent Park  43.654260 -79.360636  
14                           Ryerson, Garden District  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Adelaide, King, Richmond  43.650571 -79.384568  
19  Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752  
20           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23                Forest Hill North, Forest Hill West  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     Harbord, University of Toronto  43.662696 -79.400049  
26          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049  
27  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420  
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
29             First Canadian Place, Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                       Dovercourt Village, Dufferin  43.669005 -79.442259  
32                           Little Portugal, Trinity  43.64792

In [51]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.18.1-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  25.12 MB/s
geopy-1.18.1-p 100% |################################| Time: 0:00:00  36.38 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  55.00 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  31.07 MB/s
vincent-0.4.4- 100% |###################

<h4>Create map of Toronto with neighbourhoods superimposed on top</h4>

In [61]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[43.6532, -79.3832], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_tb['Latitude'], df_tb['Longitude'], df_tb['Borough'], df_tb['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h4>Define Foursquare Credentials and Version</h4>

In [59]:
CLIENT_ID = 'VANBONR4EFDLXUOXYA3XQHRG4DN5TUJGF3U5ZUEVPMDP4J2L' # your Foursquare ID
CLIENT_SECRET = '3HOIYMNTA5TQEJCQ2XMEEDTDRII0F0RTGDJVFEL1COJWEFUT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET: ' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VANBONR4EFDLXUOXYA3XQHRG4DN5TUJGF3U5ZUEVPMDP4J2L
CLIENT_SECRET: 3HOIYMNTA5TQEJCQ2XMEEDTDRII0F0RTGDJVFEL1COJWEFUT


<h4>Test run with one neighbourhood: M4E, East Toronto, The Beaches, (43.676357, -79.293031)</h4>

In [66]:
df_tb.reset_index(drop=True, inplace=True)

In [68]:
df_tb.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [72]:
neighborhood_latitude = df_tb.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_tb.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_tb.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [75]:
radius = 500
LIMIT = 100

In [77]:
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(\
CLIENT_ID, CLIENT_SECRET, VERSION, neighborhood_latitude, neighborhood_longitude, radius, LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=VANBONR4EFDLXUOXYA3XQHRG4DN5TUJGF3U5ZUEVPMDP4J2L&client_secret=3HOIYMNTA5TQEJCQ2XMEEDTDRII0F0RTGDJVFEL1COJWEFUT&v=20180605&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

In [78]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c3ec8c4db04f57d5569d103'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4b8daea1f964a520480833e3-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/nightlife/pub_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d11b941735',
         'name': 'Pub',
         'pluralName': 'Pubs',
         'primary': True,
         'shortName': 'Pub'}],
       'id': '4b8daea1f964a520480833e3',
       'location': {'address': '676 Kingston Rd.',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'at Main St.',
        'distance': 460,
        'formattedAddress': ['676 Kingston Rd. (at Main St.)',
         'Toronto ON M4E 1R4',
         'Canada'],
        'labeledLatLngs': [{'label': 'display',
 

In [79]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [80]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Grover Pub and Grub                   Pub  43.679181 -79.297215
1            Starbucks           Coffee Shop  43.678798 -79.298045
2        Upper Beaches          Neighborhood  43.680563 -79.292869
3      Beaches Fitness  Gym / Fitness Center  43.680319 -79.290991
4           Dip 'n Sip           Coffee Shop  43.678897 -79.297745

In [81]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


<h3>Explore the 4 boroughs of Toronto</h3>

In [82]:
# creating a function to repeat the same process for all the neighborhoods in Manhattan

In [83]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [ ]:
# call the above function for each neighborhood

In [84]:
LIMIT = 100
toronto_venues = getNearbyVenues(names=df_tb['Neighborhood'],
                                   latitudes=df_tb['Latitude'],
                                   longitudes=df_tb['Longitude']
                                  )


The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

In [85]:
print(toronto_venues.shape)
toronto_venues.head()

(1707, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                 Venue  Venue Latitude  Venue Longitude        Venue Category  
0  Grover Pub and Grub       43.679181       -79.297215                   Pub  
1            Starbucks       43.678798       -79.298045           Coffee Shop  
2        Upper Beaches       43.680563       -79.292869          Neighborhood  
3      Beaches Fitness       43.680319       -79.290991  Gym / Fitness Center  
4           Dip 'n Sip       43.678897       -79.297745           Coffee Shop

In [89]:
# checking how many venues were returned per venue
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            54   
Brockton, Exhibition Place, Parkdale Village                           19   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            47   
Central Bay Street                                                     82   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   87   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                       10   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         34   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              49   
High Park, The Junction South                                          23   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               66   
Moore Park, Summerhill East                                             4   
North Toronto West                                                     21   
Parkdale, Roncesvalles                                                 16   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     40   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   
Studio District                                                        39   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         20   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             54   
Brockton, Exhibition Place, Parkdale Village                            19   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             47   
Central Bay Street                                                      82   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                15   
Church and Wellesley                                                    87   
Commerce Court, Victoria Hotel     

In [88]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 238 uniques categories.


<h3>Analyzing each neighbourhood</h3>

In [90]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0                 0              0          0           0           0       0   
1                 0              0          0           0           0       0   
2                 0              0          0           0           0       0   
3                 0              0          0           0           0       0   
4                 0              0          0           0           0       0   

   Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  Beer Bar  \
0     0    0                 0                   0      0         0   
1     0    0                 0                   0      0         0   
2     0    0                 0                   0      0         0   
3     0    0                 0                   0      0         0   
4     0    0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

   Café  Cajun / Creole Restaurant  Caribbean Restaurant  Cheese Shop  \
0     0                          

In [91]:
toronto_onehot.shape

(1707, 238)

In [92]:
# Grouping rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [93]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.058824   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.012195   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011494   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.000000   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.000000   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015152   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.047619   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.010000   
32                    Stn A PO Boxes 25 The Esplanade     0.010526   
33                                    Studio District     0.025641   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011494           0.011494  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.0

In [94]:
toronto_grouped.shape

(38, 238)

In [95]:
# print the top 5 venues for each neighbourhood
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3      Thai Restaurant  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1          Restaurant  0.06
2        Cocktail Bar  0.06
3         Cheese Shop  0.04
4  Italian Restaurant  0.04


----Brockton, Exhibition Place, Parkdale Village----
            venue  freq
0  Breakfast Spot  0.11
1            Café  0.11
2     Coffee Shop  0.11
3    Climbing Gym  0.05
4             Bar  0.05


----Business Reply Mail Processing Centre 969 Eastern----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2        Comic Shop  0.06
3              Park  0.06
4  Recording Studio  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  0.14
1   Airport Service

<h3>Putting all the generated information into a Pandas Dataframe for further processing </h3>

In [96]:
# sorting venues function
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [97]:
# creating new dataframe with top 10 venues
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant            Steakhouse   
1            Restaurant          Cocktail Bar           Cheese Shop   
2                  Café        Breakfast Spot                   Bar   
3         Auto Workshop                  Park           Pizza Place   
4      Airport Terminal       Airport Service              Boutique   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant                 Hotel            Restaurant   
1                Bakery    Italian Restaurant              Beer Bar   
2               Stadium    Italian Restaurant         Burrito Place   
3      Recording Studio            Restaurant         Burrito Place   
4      Sculpture Garden                 Plane         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0      Asian Restaurant                   Bar                     Gym  
1            Steakhouse    Seafood Restaurant                     Pub  
2          Climbing Gym                Office  Furniture / Home Store  
3               Brewery            Skate Park              Smoke Shop  
4          Airport Gate    Airport Food Court                 Airport

<h3>Cluster Neighbourhoods</h3>

In [98]:
# k-means 
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 1, 2, 2, 2, 2, 2, 2], dtype=int32)

In [99]:
# creating a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood
toronto_merged = df_tb

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               2           Coffee Shop  Gym / Fitness Center   
1 -79.352188               2      Greek Restaurant           Coffee Shop   
2 -79.315572               2                  Park      Sushi Restaurant   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               2              Bus Line                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub         Women's Store        Discount Store   
1        Ice Cream Shop             Bookstore    Italian Restaurant   
2            Board Shop               Brewery          Burger Joint   
3                Bakery    Italian Restaurant   American Restaurant   
4           Swim School    Dim Sum Restaurant         Women's Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Fast Food Restaurant        Farmers Market    Falafel Restaurant   
1           Yoga Studio        Cosmetics Shop               Brewery   
2        Sandwich Place         Burrito Place                   Pub   
3           Yoga Studio       Coworking Space    Seafood Restaurant   
4  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space   Ethiopian Restaurant  
1       Bubble Tea Shop             Restaurant  
2             Pet Store            Pizza Place  
3        Sandwich Place            Cheese Shop  
4           Event Space   Ethiopian Restaurant

In [101]:
# visualising the results
# create map
map_clusters = folium.Map(location=[43.6532, -79.3832], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<h3>Examining the resultant clusters</h3>

In [102]:
# cluster 1
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
24   Central Toronto               0           Coffee Shop   
27  Downtown Toronto               0        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24                  Café        Sandwich Place           Pizza Place   
27      Airport Terminal       Airport Service              Boutique   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
24        History Museum     Indian Restaurant              Pharmacy   
27      Sculpture Garden                 Plane         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24        Cosmetics Shop                   Pub           Burger Joint  
27          Airport Gate    Airport Food Court                Airport

In [103]:
# cluster 2
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
3       East Toronto               1                  Café   
12  Downtown Toronto               1   Japanese Restaurant   
22   Central Toronto               1                Garden   
36      West Toronto               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3            Coffee Shop                Bakery    Italian Restaurant   
12      Sushi Restaurant           Coffee Shop               Gay Bar   
22                  Pool           Music Venue         Women's Store   
36           Pizza Place                  Café      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3    American Restaurant           Yoga Studio       Coworking Space   
12            Restaurant          Burger Joint       Bubble Tea Shop   
22        Farmers Market    Falafel Restaurant           Event Space   
36                 Diner    Italian Restaurant          Gourmet Shop   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
3     Seafood Restaurant        Sandwich Place                  Cheese Shop  
12                  Café  Fast Food Restaurant                  Men's Store  
22  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
36             Bookstore                  Food                          Bar

In [104]:
# cluster 3
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               2           Coffee Shop   
1       East Toronto               2      Greek Restaurant   
2       East Toronto               2                  Park   
4    Central Toronto               2              Bus Line   
5    Central Toronto               2                  Park   
6    Central Toronto               2   Sporting Goods Shop   
7    Central Toronto               2           Pizza Place   
8    Central Toronto               2            Playground   
9    Central Toronto               2                   Pub   
10  Downtown Toronto               2                  Park   
11  Downtown Toronto               2           Coffee Shop   
13  Downtown Toronto               2           Coffee Shop   
14  Downtown Toronto               2           Coffee Shop   
15  Downtown Toronto               2           Coffee Shop   
16  Downtown Toronto               2           Coffee Shop   
18  Downtown Toronto               2           Coffee Shop   
19  Downtown Toronto               2           Coffee Shop   
20  Downtown Toronto               2           Coffee Shop   
21  Downtown Toronto               2           Coffee Shop   
23   Central Toronto               2              Bus Line   
25  Downtown Toronto               2                  Café   
26  Downtown Toronto               2                   Bar   
29  Downtown Toronto               2                  Café   
30  Downtown Toronto               2                  Café   
31      West Toronto               2              Pharmacy   
32      West Toronto               2                   Bar   
33      West Toronto               2           Coffee Shop   
34      West Toronto               2    Mexican Restaurant   
35      West Toronto               2        Breakfast Spot   
37      East Toronto               2           Yoga Studio   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0   Gym / Fitness Center                            Pub         Women's Store   
1            Coffee Shop                 Ice Cream Shop             Bookstore   
2       Sushi Restaurant                     Board Shop               Brewery   
4                   Park                    Swim School    Dim Sum Restaurant   
5                  Hotel                   Burger Joint     Food & Drink Shop   
6            Coffee Shop                 Clothing Store    Salon / Barbershop   
7           Dessert Shop                 Sandwich Place    Italian Restaurant   
8                    Gym                   Tennis Court                  Park   
9            Coffee Shop             Light Rail Station           Supermarket   
10            Playground                          Trail                 Diner   
11            Restaurant             Italian Restaurant           Pizza Place   
13                  Park                           Café                Bakery   
14        Clothing Store                           Café        Cosmetics Shop   
15            Restaurant                           Café                 Hotel   
16            Restaurant                   Cocktail Bar           Cheese Shop   
18                  Café                Thai Restaurant            Steakhouse   
19              Aquarium                          Hotel           Pizza Place   
20                 Hotel                           Café   American Restaurant   
21                  Café                          Hotel            Restaurant   
23                 Trail               Sushi Restaurant         Jewelry Store   
25           Coffee Shop                         Bakery             Bookstore   
26                  Café  Vegetarian / Vegan Restaurant   Dumpling Restaurant   
29           Coffee Shop                          Hotel            Restaurant   
30         Grocery Store                           Park     Convenience Store   
31           Supermarket                 Discount Store          

In [105]:
# cluster 4
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               3           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café    Italian Restaurant                   Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17          Burger Joint    Chinese Restaurant                   Spa   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17        Sandwich Place    Falafel Restaurant      Indian Restaurant

In [106]:
# cluster 5
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]



Borough  Cluster Labels 1st Most Common Venue  \
28  Downtown Toronto               4           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Restaurant                  Café    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28    Seafood Restaurant                   Pub                 Hotel   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28              Beer Bar          Cocktail Bar    Japanese Restaurant